salt/channel/client.py

```diff
@@ -0,0 +1,555 @@
+        Send a load across the wire in cleartext
+
+        :param dict load: A load to send across the wire
+        :param int tries: The number of times to make before failure
+        :param int timeout: The number of seconds on a response before failing
+        """
+        ret = yield self.transport.send(
+            self._package_load(load),
+            timeout=timeout,
+            tries=tries,
+        )
+
+        raise salt.ext.tornado.gen.Return(ret)
+
+    @salt.ext.tornado.gen.coroutine
+    def send(self, load, tries=3, timeout=60, raw=False):
+        """
+        Send a request, return a future which will complete when we send the message
+        """
+        if self.crypt == "clear":
+            ret = yield self._uncrypted_transfer(load, tries=tries, timeout=timeout)
+        else:
+            ret = yield self._crypted_transfer(
+                load, tries=tries, timeout=timeout, raw=raw
+            )
+        raise salt.ext.tornado.gen.Return(ret)
+
+    def close(self):
+        """
+        Since the message_client creates sockets and assigns them to the IOLoop we have to
+        specifically destroy them, since we aren't the only ones with references to the FDs
+        """
+        if self._closing:
+            return
+        log.debug("Closing %s instance", self.__class__.__name__)
+        self._closing = True
+        self.transport.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+
+class AsyncPubChannel:
+    """
+    Factory class to create subscription channels to the master's Publisher
+    """
+
+    async_methods = [
+        "connect",
+        "_decode_messages",
+    ]
+    close_methods = [
+        "close",
+    ]
+
+    @classmethod
+    def factory(cls, opts, **kwargs):
+        import salt.ext.tornado.ioloop
+        import salt.crypt
+
+        # Default to ZeroMQ for now
+        ttype = "zeromq"
+
+        # determine the ttype
+        if "transport" in opts:
+            ttype = opts["transport"]
+        elif "transport" in opts.get("pillar", {}).get("master", {}):
+            ttype = opts["pillar"]["master"]["transport"]
+
+        if "master_uri" not in opts and "master_uri" in kwargs:
+            opts["master_uri"] = kwargs["master_uri"]
+
+        # switch on available ttypes
+        if ttype == "detect":
+            opts["detect_mode"] = True
+            log.info("Transport is set to detect; using %s", ttype)
+
+        io_loop = kwargs.get("io_loop")
+        if io_loop is None:
+            io_loop = salt.ext.tornado.ioloop.IOLoop.current()
+
+        auth = salt.crypt.AsyncAuth(opts, io_loop=io_loop)
+        transport = salt.transport.publish_client(opts, io_loop)
+        return cls(opts, transport, auth, io_loop)
+
+    def __init__(self, opts, transport, auth, io_loop=None):
+        self.opts = opts
+        self.io_loop = io_loop
+        self.auth = auth
+        self.tok = self.auth.gen_token(b"salt")
```

Comment:
**comment** I see this isn't code that originated here. `self.token` would be clearer - from what I can tell that wouldn't be a huge problem - the payload would probably still have to be `"tok": self.token,` but... :shrug:

Reply:
Makes sense to me.